software/infnoise.8
.\"                                      Hey, EMACS: -*- nroff -*-
.TH INFNOISE 8 "June 25 2018"
.\" Please adjust this date whenever revising the manpage.
.SH NAME
infnoise \- Infinite Noise TRNG driver
.SH SYNOPSIS
.B infnoise \-\-help
.br
.B infnoise \-\-version
.br
.B infnoise \-\-list\-devices
.br
.B infnoise
.RB [ \-\-debug ]
.RB [ \-\-dev\-random ]
.RB [ \-\-raw ]
.RB [ \-\-multiplier
.RI < value >]
.RB [ \-\-no\-output ]
.RB [ \-\-pidfile
.RI < file >]
.RB [ \-\-daemon ]
.RB [ \-\-serial
.RI < serial >]
.SH DESCRIPTION
.B infnoise
provides access to the Infinite Noise True Random Number Generator and
allows the data it generates to be fed into the system's random number
generator.
.PP
.B infnoise \-\-help
shows a short summary of the options.
.PP
.B infnoise \-\-version
displays the program's version information.
.PP
.B infnoise \-\-list\-devices
lists all the supported TRNG devices present on the system.
.PP
.B infnoise
reads random data from a TRNG and outputs filtered random data to its
standard output. The various options control the program's behaviour.
.SH OPTIONS
.TP
.B \-\-debug
measures the quality of the data obtained from the TRNG. It is
typically used with \-\-no\-output as a sanity test.
.TP
.B \-\-dev\-random
adds the filtered data to the system's entropy pool instead of sending
it to its standard output.
.TP
.B \-\-raw
disables output whitening.
.TP
.BR \-\-multiplier <\fIvalue\fP>
writes 256 bits ×
.I value
for every 512 bits written to the Keccak sponge; the default of 0
means to write all the available entropy, without multiplying it.
.TP
.B \-\-no\-output
disables output of random data.
.TP
.BR \-\-pidfile <\fIfile\fP>
stores the process' identifier in
.IR file .
.TP
.B \-\-daemon
starts the program as a daemon.
.TP
.BR \-\-serial <\fIserial\fP>
uses the device matching the specified
.IR serial .


CHANGELOG.md
# Change Log

## [0.3.0](https://github.com/13-37-org/infnoise/tree/0.3.0) (2018-10-09)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/0.2.6...0.3.0)

**Merged pull requests:**

- merge libinfnoise branch [\#12](https://github.com/13-37-org/infnoise/pull/12) ([manuel-domke](https://github.com/manuel-domke))
- Fix apt-key add command in README [\#11](https://github.com/13-37-org/infnoise/pull/11) ([xenomachina](https://github.com/xenomachina))
- Fix a couple of spelling errors [\#10](https://github.com/13-37-org/infnoise/pull/10) ([skitt](https://github.com/skitt))

## [0.2.6](https://github.com/13-37-org/infnoise/tree/0.2.6) (2018-05-02)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/0.2.5...0.2.6)

**Closed issues:**

- systemd service can't be disabled [\#7](https://github.com/13-37-org/infnoise/issues/7)

## [0.2.5](https://github.com/13-37-org/infnoise/tree/0.2.5) (2018-04-09)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/0.2.4...0.2.5)

## [0.2.4](https://github.com/13-37-org/infnoise/tree/0.2.4) (2018-04-02)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/0.2.3...0.2.4)

## [0.2.3](https://github.com/13-37-org/infnoise/tree/0.2.3) (2018-03-02)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/0.2.2...0.2.3)

## [0.2.2](https://github.com/13-37-org/infnoise/tree/0.2.2) (2018-03-01)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/0.2.0...0.2.2)

## [0.2.0](https://github.com/13-37-org/infnoise/tree/0.2.0) (2018-03-01)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/0.2.1...0.2.0)

## [0.2.1](https://github.com/13-37-org/infnoise/tree/0.2.1) (2018-03-01)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/v0.2.1...0.2.1)

## [v0.2.1](https://github.com/13-37-org/infnoise/tree/v0.2.1) (2018-03-01)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/v0.2.0...v0.2.1)

**Merged pull requests:**

- Version number [\#5](https://github.com/13-37-org/infnoise/pull/5) ([manuel-domke](https://github.com/manuel-domke))
- Correct CS spelling [\#4](https://github.com/13-37-org/infnoise/pull/4) ([darrellrossman](https://github.com/darrellrossman))

## [v0.2.0](https://github.com/13-37-org/infnoise/tree/v0.2.0) (2017-12-31)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/v0.1.2...v0.2.0)

**Merged pull requests:**

- merged captainpete's pull request \#28 from waywardgeek/infnoise [\#3](https://github.com/13-37-org/infnoise/pull/3) ([manuel-domke](https://github.com/manuel-domke))
- merged "increased timeout and options struct" pull request from runema [\#2](https://github.com/13-37-org/infnoise/pull/2) ([manuel-domke](https://github.com/manuel-domke))

## [v0.1.2](https://github.com/13-37-org/infnoise/tree/v0.1.2) (2017-11-13)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/0.1.0...v0.1.2)

**Merged pull requests:**

- updated readme [\#1](https://github.com/13-37-org/infnoise/pull/1) ([manuel-domke](https://github.com/manuel-domke))

## [0.1.0](https://github.com/13-37-org/infnoise/tree/0.1.0) (2017-08-29)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/0.1.1...0.1.0)

## [0.1.1](https://github.com/13-37-org/infnoise/tree/0.1.1) (2017-08-29)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/Fourth-OshPark-Prototype...0.1.1)

## [Fourth-OshPark-Prototype](https://github.com/13-37-org/infnoise/tree/Fourth-OshPark-Prototype) (2014-11-02)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/Third-OshPark-Prototype...Fourth-OshPark-Prototype)

## [Third-OshPark-Prototype](https://github.com/13-37-org/infnoise/tree/Third-OshPark-Prototype) (2014-10-15)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/Second-OshPark-Prototype...Third-OshPark-Prototype)

## [Second-OshPark-Prototype](https://github.com/13-37-org/infnoise/tree/Second-OshPark-Prototype) (2014-10-12)
[Full Changelog](https://github.com/13-37-org/infnoise/compare/First-OshPark-Prototype...Second-OshPark-Prototype)

## [First-OshPark-Prototype](https://github.com/13-37-org/infnoise/tree/First-OshPark-Prototype) (2014-10-10)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


software/infnoise.c
/* Driver for the Infinite Noise Multiplier USB stick */

// Required to include clock_gettime
#define _POSIX_C_SOURCE 200809L

#include <stdint.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#if defined(__OpenBSD__) || defined(__NetBSD__) || defined(__DragonFly__) || defined(__FreeBSD__)
#include <fcntl.h>
#endif

#include <string.h>
#include <time.h>
#include <sys/types.h>
#include <ftdi.h> // requires <sys/types.h>
#include <getopt.h>
#include "infnoise.h"
#include "libinfnoise.h"

static void initOpts(struct opt_struct *opts) {
    opts->outputMultiplier = 0u;
    opts->daemon = false;
    opts->debug = false;
    opts->devRandom = false;
    opts->noOutput = false;
    opts->listDevices = false;
    opts->raw = false;
    opts->version = false;
    opts->help = false;
    opts->none = false;
    opts->pidFileName =
    opts->serial = NULL;
}

// getopt_long(3) options descriptor
static struct option longopts[] = {
        {"raw",          no_argument,       NULL, 'r'},
        {"debug",        no_argument,       NULL, 'D'},
        {"dev-random",   no_argument,       NULL, 'R'},
        {"no-output",    no_argument,       NULL, 'n'},
        {"multiplier",   required_argument, NULL, 'm'},
        {"pidfile",      required_argument, NULL, 'p'},
        {"serial",       required_argument, NULL, 's'},
        {"daemon",       no_argument,       NULL, 'd'},
        {"list-devices", no_argument,       NULL, 'l'},
        {"version",      no_argument,       NULL, 'v'},
        {"help",         no_argument,       NULL, 'h'},
        {NULL,           0,                 NULL, 0}};

// Write the bytes to either stdout, or /dev/random.
bool outputBytes(uint8_t *bytes, uint32_t length, uint32_t entropy, bool writeDevRandom, char **message) {
    if (!writeDevRandom) {
        if (fwrite(bytes, 1, length, stdout) != length) {
            *message = "Unable to write output from Infinite Noise Multiplier";
            return false;
        }
    } else {
#if defined(__OpenBSD__) || defined(__NetBSD__) || defined(__DragonFly__) || defined(__FreeBSD__)
        // quell compiler warning about unused variable.
        static int devRandomFD = -1;
        (void) entropy;

        if (devRandomFD < 0)
            devRandomFD = open("/dev/random", O_WRONLY);
        if (devRandomFD < 0) {
            *message = "Unable to open random(4)";
            return false;
        };
        // we are not trapping EINT and EAGAIN; as the random(4) driver seems
        // to not treat partial writes as not an error. So we think that comparing
        // to length is fine.
        //
        if (write(devRandomFD, bytes, length) != length) {
            *message = "Unable to write output from Infinite Noise Multiplier to random(4)";
            return false;
        }
#endif
#if defined(__APPLE__)
        *message = "macOS doesn't support writes to entropy pool";
        entropy = 0; // suppress warning
        return false;
#endif
#ifdef LINUX
        inmWaitForPoolToHaveRoom();
        inmWriteEntropyToPool(bytes, length, entropy);
#endif
    }
    return true;
}

int main(int argc, char **argv) {
    struct infnoise_context context;
    struct opt_struct opts;
    int ch;
    bool multiplierAssigned = false;

    initOpts(&opts);

    // Process arguments
    while ((ch = getopt_long(argc, argv, "rDRnm:p:s:dlvh", longopts, NULL)) !=
           -1) {
        switch (ch) {
            case 'r':
                opts.raw = true;
                break;
            case 'D':
                opts.debug = true;
                break;
            case 'R':
                opts.devRandom = true;
                break;
            case 'n':
                opts.noOutput = true;
                break;
            case 'm':
                multiplierAssigned = true;
                int tmpOutputMult = atoi(optarg);
                if (tmpOutputMult < 0) {
                    fputs("Multiplier must be >= 0\n", stderr);
                    return 1;
                }
                opts.outputMultiplier = tmpOutputMult;
                break;
            case 'p':
                opts.pidFileName = optarg;
                if (opts.pidFileName == NULL || !strcmp("", opts.pidFileName)) {
                    fputs("--pidfile without file name\n", stderr);
                    return 1;
                }
                break;
            case 's':
                opts.serial = optarg;
                if (opts.serial == NULL || !strcmp("", opts.serial)) {
                    fputs("--serial without value\n", stderr);
                    return 1;
                }
                break;
            case 'd':
                opts.daemon = true;
                break;
            case 'l':
                opts.listDevices = true;
                break;
            case 'v':
                opts.version = true;
                break;
            case 'h':
                opts.help = true;
                break;
            default:
                opts.help = true;
                opts.none = true;
        }
    }

    if (opts.help) {
        fputs("Usage: infnoise [options]\n"
              "Options are:\n"
              "    -D, --debug - turn on some debug output\n"
              "    -R, --dev-random - write entropy to /dev/random instead of "
              "stdout\n"
              "    -r, --raw - do not whiten the output\n"
              "    -m, --multiplier <value> - write 256 bits * value for each 512 bits written to\n"
              "      the Keccak sponge.  Default of 0 means write all the entropy.\n"
              "    -n, --no-output - do not write random output data\n"
              "    -p, --pidfile <file> - write process ID to file\n"
              "    -d, --daemon - run in the background\n"
              "    -s, --serial <serial> - use specified device\n"
              "    -l, --list-devices - list available devices\n"
              "    -v, --version - show version information\n"
              "    -h, --help - this help output\n",
              stdout);
        if (opts.none) {
            return 1;
        } else {
            return 0;
        }
    }

    // read environment variables, not overriding command line options
    if (opts.serial == NULL) {
        if (getenv("INFNOISE_SERIAL") != NULL) {
            opts.serial = getenv("INFNOISE_SERIAL");
        }
    }

    if (!opts.debug) {
        if (getenv("INFNOISE_DEBUG") != NULL) {
            if (!strcmp("true", getenv("INFNOISE_DEBUG"))) {
                opts.debug = true;
            }
        }
    }

    if (!multiplierAssigned) {
        if (getenv("INFNOISE_MULTIPLIER") != NULL) {
            int tmpOutputMult = atoi(getenv("INFNOISE_MULTIPLIER"));
            if (tmpOutputMult < 0) {
                fputs("Multiplier must be >= 0\n", stderr);
                return 1;
            }
            multiplierAssigned = true;
            opts.outputMultiplier = tmpOutputMult;
        }
    }

    if (!multiplierAssigned && opts.devRandom) {
        opts.outputMultiplier = 2u; // Don't throw away entropy when writing to /dev/random unless told to do so
    }

    if (opts.version) {
        printf("GIT VERSION - %s\n", GIT_VERSION);
        printf("GIT COMMIT  - %s\n", GIT_COMMIT);
        printf("GIT DATE    - %s\n", GIT_DATE);
        return 0;
    }

    if (opts.listDevices) {
        devlist_node devlist = listUSBDevices(&context.message);
        if (devlist == NULL) {
            fprintf(stderr, "Error: %s\n", context.message);
            return 1;
        }
        devlist_node curdev = NULL;
        uint8_t i = 0;
        for (curdev = devlist; curdev != NULL; i++) {
            printf("ID: %i, Manufacturer: %s, Description: %s, Serial: %s\n", curdev->id, curdev->manufacturer,
                   curdev->description, curdev->serial);
            curdev = curdev->next;
        }
        return 0;
    }

    if (opts.devRandom) {
#ifdef LINUX
        inmWriteEntropyStart(BUFLEN / 8u, opts.debug); // TODO: create method in libinfnoise.h for this?
        // also TODO: check superUser in this mode (it will fail silently if not :-/)
#endif
#if defined(__OpenBSD__) || defined(__NetBSD__) || defined(__DragonFly__) || defined(__FreeBSD__)
        int devRandomFD = open("/dev/random", O_WRONLY);
        if (devRandomFD < 0) {
            fprintf(stderr, "Unable to open /dev/random\n");
            exit(1);
        }
        close(devRandomFD);
#endif
#if defined(__APPLE__)
        context.message = "dev/random not supported on macOS";
        fprintf(stderr, "Error: %s\n", context.message);
        return 1;
#endif
    }

    // Optionally run in the background and optionally write a PID-file
    startDaemon(&opts);

    // initialize USB device, health check and Keccak state (see libinfnoise)
    if (!initInfnoise(&context, opts.serial, !opts.raw, opts.debug)) {
        fprintf(stderr, "Error: %s\n", context.message);
        return 1; // ERROR
    }
    context.errorFlag = false;

    // calculate output size based on the parameters:
    uint64_t resultSize;
    if (opts.outputMultiplier <= 2 || opts.raw) {
        resultSize = 64u;
    } else {
        resultSize = 128u;
    }
    //fprintf(stderr, "resultsize: %lu\n", resultSize);

    // endless loop
    uint64_t totalBytesWritten = 0u;
    while (true) {
        uint8_t result[resultSize];
        uint64_t bytesWritten = readData(&context, result, opts.raw, opts.outputMultiplier);
        totalBytesWritten += bytesWritten;

        if (context.errorFlag) {
            fprintf(stderr, "Error: %s\n", context.message);
            return 1;
        }

        if (!opts.noOutput) {
            if (!outputBytes(result, bytesWritten, context.entropyThisTime, opts.devRandom,
                             &context.message)) {
                fprintf(stderr, "Error: %s\n", context.message);
                return 1;
            };
        }

        if (opts.debug &&
            (1u << 20u) * (totalBytesWritten / (1u << 20u)) > (1u << 20u) * (totalBytesWritten + bytesWritten / (1u << 20u))) {
            fprintf(stderr, "Output %lu bytes\n", (unsigned long) totalBytesWritten);
        }
    }
}
